### PR TITLE
FEAT: create KB from followup or task

### PR DESCRIPTION
**Pull request by Rom1-B**
Allows to create a kb from followup or task, as it already exists in solution.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !22387

### CODE REVIEW

**Comment by trasher** [comment]
This should target master

File: front/change.form.php [hunk 1]
```diff
@@ -139,16 +139,29 @@
...
+      } else if (isset($_GET['_task_to_kb'])) {
+         $url .= '&task_id=' . $_GET['_task_to_kb'];
+         $handled = true;
+      }
+      if ($handled) {
+         Ajax::createIframeModalWindow(
+            'savetokb',
+            $url,
+            [
+               'title'         => __('Save solution to the knowledge base'),
```
Comment (trasher) [inline front/change.form.php:159]: As far as I understand, there are 3 possible cases; title here is OK for one only

**Review by trasher — COMMENTED** [review]
Maybe all the logic inside front files maybe moved in class

File: inc/knowbaseitem.class.php [hunk 2]
```diff
@@ -699,15 +699,30 @@ function showForm($ID, $options = []) {
...
-               $this->fields['answer'] = $solution->getField('content');
+               if (isset($options['fw_id'])) {
+                  $fup = new ITILFollowup();
+                  $fup->getFromDBByCrit([
+                     'id'           => $options['fw_id'],
+                     'itemtype'     => $item->getType(),
+                     'items_id'     => $item->getID()
+                  ]);
+                  $this->fields['answer'] = $fup->getField('content');
+               } else if (isset($options['task_id'])) {
```
Comment (cedric-anne) [inline inc/knowbaseitem.class.php:710]:
```suggestion
               } else if (isset($options['tasks_id'])) {
```

File: inc/knowbaseitem.class.php [hunk 3]
```diff
@@ -699,15 +699,30 @@ function showForm($ID, $options = []) {
...
-               $solution = new ITILSolution();
-               $solution->getFromDBByCrit([
-                  'itemtype'     => $item->getType(),
-                  'items_id'     => $item->getID(),
-                  [
-                     'NOT' => ['status'       => CommonITILValidation::REFUSED]
-                  ]
-               ]);
-               $this->fields['answer'] = $solution->getField('content');
+               if (isset($options['fw_id'])) {
```
Comment (cedric-anne) [inline inc/knowbaseitem.class.php:702]:
```suggestion
               if (isset($options['itilfollowups_id'])) {
```

File: inc/knowbaseitem.class.php [hunk 4]
```diff
@@ -699,15 +699,30 @@ function showForm($ID, $options = []) {
...
+                  $fup->getFromDBByCrit([
+                     'id'           => $options['fw_id'],
+                     'itemtype'     => $item->getType(),
+                     'items_id'     => $item->getID()
+                  ]);
+                  $this->fields['answer'] = $fup->getField('content');
+               } else if (isset($options['task_id'])) {
+                  $tasktype = $item->getType().'Task';
+                  $task = new $tasktype;
+                  $task->getFromDB($options['task_id']);
```
Comment (cedric-anne) [inline inc/knowbaseitem.class.php:713]:
```suggestion
                  $task->getFromDB($options['tasks_id']);
```

File: front/itilfollowup.form.php [hunk 5]
```diff
@@ -80,7 +84,45 @@
...
+}
+
+if (null == $redirect) {
+   Html::back();
+} else {
+   Html::redirect($redirect);
 }
 
-Html::displayErrorAndDie('Lost');
+Html::displayErrorAndDie('Lost');
```
Comment (orthagh) [inline front/itilfollowup.form.php:128]:
```suggestion
Html::displayErrorAndDie('Lost');

```

**Review by orthagh — CHANGES_REQUESTED** [review]

**Comment by orthagh** [comment]
An aditionnal PR need to be done to redo the UI part in modern-UI

**Review by orthagh — APPROVED** [review]

**Review by cedric-anne — APPROVED** [review]

**Comment by cedric-anne** [comment]
Replaced by #9620

**Pull request closed without merging**